crates/sui-edge-proxy/src/handlers.rs

```diff
@@ -205,5 +212,30 @@ async fn proxy_request(
             resp.headers_mut().insert(name, value);
         }
     }
+
+    // Only log read requests, check sampling
+    if matches!(peer_type, PeerRole::Read) {
+        let rate = state.logging_config.read_request_sample_rate;
+        if rand::thread_rng().gen::<f64>() < rate {
+            let log_entry = format!(
+                "HEADERS: {:?}\nBODY: {:?}\nPEER_TYPE: {:?}\n\n",
+                parts.headers, body_bytes, peer_type
+            );
+
+            let _guard = state.log_file_lock.lock().await;
+            if let Err(e) = std::fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(&state.logging_config.log_file_path)
```

Comment:
Can we open the file once and store it in the `AppState` as an `Option<Mutex<File>>` instead of re-opening it every time we get a request?

I would even be tempted to abstract this into a `Logger` type that hides all this -- the handler just knows that if it sees a Read, it needs to send the info to the logger, the logger can then decide what to do (discard because of sampling, write to a buffer, flush the file, etc).

Ideally you wouldn't even do the flush on the critical path of `proxy_request` -- that would just write to memory, and some other job would flush some prefix of that buffer off the critical path. I'd be happy to wait to see if that optimisation is necessary because it's a bit more involved though.

Reply:
very valid point while taking a step back, John and I were talking about using grafana / loki for the logging instead with a couple of advantages:
- I do not have to manually mount and manage log files on k8 pods, instead the grafana infra is set in production
- the filter / search capability provided by grafana also comes handy

the only "downside" is that, I will need to write another script to export the logs from grafana to a log file, which is straight-forward, so I will go that direction and changing the pr here.

Reply:
Even better!